server/app/services/question/ActiveAndDraftQuestions.java

```diff
@@ -24,39 +33,95 @@ public final class ActiveAndDraftQuestions {
     }
-    versionedByName = versionedByNameBuilder.build();
-    for (String questionName : activeNames.keySet()) {
-      if (draft.getTombstonedQuestionNames().contains(questionName)) {
-        deletionStatusBuilder.put(questionName, DeletionStatus.PENDING_DELETION);
-      } else if (isNotDeletable(active, draft, activeNames, draftNames, questionName)) {
-        deletionStatusBuilder.put(questionName, DeletionStatus.NOT_DELETABLE);
-      } else {
-        deletionStatusBuilder.put(questionName, DeletionStatus.DELETABLE);
+
+    deletionStatusByName =
+        activeNames.keySet().stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    questionName -> questionName,
+                    questionName -> {
+                      if (draft.getTombstonedQuestionNames().contains(questionName)) {
+                        return DeletionStatus.PENDING_DELETION;
+                      } else if (isNotDeletable(
+                          active, draft, activeNames, draftNames, questionName)) {
+                        return DeletionStatus.NOT_DELETABLE;
+                      } else {
+                        return DeletionStatus.DELETABLE;
+                      }
+                    }));
+  }
+
+  /**
+   * Inspects the provided version and returns a map who's key is the question name and value is a
+   * set of programs that reference the given question in this version.
+   */
+  private static ImmutableMap<String, ImmutableSet<ProgramDefinition>> buildReferencingProgramsMap(
+      Version version) {
+    // Different versions of a question can have distinct IDs while still
+    // retaining the same "name". A given program has a reference only to a specific
+    // question ID. This map allows us to easily cache the mapping from a question ID
+    // to a logical question "name".
+    ImmutableMap<Long, String> questionIdToNameLookup =
+        version.getQuestions().stream()
+            .map(Question::getQuestionDefinition)
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    QuestionDefinition::getId, QuestionDefinition::getName));
+    Map<String, Set<ProgramDefinition>> result = Maps.newHashMap();
+    for (Program program : version.getPrograms()) {
+      ImmutableList<String> programQuestionNames =
+          program.getProgramDefinition().blockDefinitions().stream()
+              .map(BlockDefinition::programQuestionDefinitions)
+              .flatMap(ImmutableList::stream)
+              .map(ProgramQuestionDefinition::id)
+              .filter(questionIdToNameLookup::containsKey)
+              .map(questionIdToNameLookup::get)
+              .collect(ImmutableList.toImmutableList());
+      for (String questionName : programQuestionNames) {
+        if (!result.containsKey(questionName)) {
+          result.put(questionName, Sets.newHashSet());
+        }
+        result.get(questionName).add(program.getProgramDefinition());
       }
```

Comment:
suggestion (non-blocking): instead of having the stream and then the loop could do `.forEach(questionName -> {...})` instead of `.collect` with the `for` loop

Reply:
Yeah, just brought it up as an option. If you this reads better to you then sgtm